README.md
[![Build Status](https://travis-ci.org/MozillaSecurity/funfuzz.svg?branch=master)](https://travis-ci.org/MozillaSecurity/funfuzz) [![codecov](https://codecov.io/gh/MozillaSecurity/funfuzz/branch/master/graph/badge.svg)](https://codecov.io/gh/MozillaSecurity/funfuzz)

This repository contains several JavaScript-based fuzzers. [jsfunfuzz](js/jsfunfuzz) tests JavaScript engines and can run in a JavaScript shell, compare_jit compares output from SpiderMonkey using different flags, while randorderfuzz throws in random tests from the mozilla-central directory into generated jsfunfuzz output.

Most of the code other than testcase generation is written in Python: restarting the program when it exits or crashes, noticing evidence of new bugs from the program's output, [reducing testcases](https://github.com/MozillaSecurity/lithium/), and [identifying when regressions were introduced](src/funfuzz/autobisectjs/README.md).


## Setup

Install the required pip packages using `pip install -r requirements.txt` (assuming you are in the funfuzz repository).

Some parts of the fuzzer will only activate if the Python scripts can find your mozilla-central tree:
```
mkdir -p ~/trees/
hg clone https://hg.mozilla.org/mozilla-central/ ~/trees/mozilla-central/
```

Some parts of the harness assume a clean **Mercurial** clone of the mozilla trees. There is insufficient testing with Git for now - please file an issue if you hit problems with Git repositories of mozilla trees.

If you want to use these scripts to compile SpiderMonkey, install the usual prerequisites for [building SpiderMonkey](https://developer.mozilla.org/en-US/docs/Mozilla/Projects/SpiderMonkey/Build_Documentation). There are [additional requirements for building with Address Sanitizer](https://developer.mozilla.org/en-US/docs/Mozilla/Testing/Firefox_and_Address_Sanitizer).

### Windows (only 64-bit supported)

1. Install [MozillaBuild](https://wiki.mozilla.org/MozillaBuild) (Using compile_shell for SpiderMonkey requires at least version 3.2).
2. Install [Git](https://git-scm.com/) to clone these funfuzz repositories.
3. Install [Debugging Tools for Windows](https://msdn.microsoft.com/en-us/windows/hardware/hh852365.aspx) to get cdb.exe and thus stacks from crashes.
4. Make sure you install at least Microsoft Visual Studio 2017 (Community Edition is recommended) as per the build instructions above in the Setup section.
5. Run `start-shell.bat` to get a MSYS shell. You can use Git by calling its absolute path, e.g. `/c/Program\ Files/Git/bin/git.exe`.
    1. Run the batch file with administrator privileges to get gflags analysis working correctly.


### Mac

1. If you encounter problems accessing the compiler, try re-running this command:

  ```xcode-select --install```

especially after updating major/minor OS versions. This sometimes manifests on Mac OS X Combo updates.

2. Install LLVM via Homebrew, to get llvm-symbolizer needed for symbolizing ASan crash stacks.

  ```
  brew install llvm
  ```


### Linux

1. To ensure your core dumps don't get mixed up when multiple instances crash at the same time, run:

  ```
  echo -n 1 | sudo tee /proc/sys/kernel/core_uses_pid
  ```
2. Install 32-bit libraries to compile 32-bit binaries:
  * Debian/Ubuntu: ```sudo apt-get install lib32z1 gcc-multilib g++-multilib```
  * Fedora: (Fedora is known to work, however the exact library names are unknown for now.)
3. Install gdb:
  * Debian/Ubuntu: ```sudo apt-get install gdb```
  * Fedora: Please ensure that all development packages are installed (see ```rpm -qa "*devel"```), and run ```yum install gdb```
4. Install clang for clang/ASan builds:
  * Debian/Ubuntu: ```sudo apt-get install clang```
  * Clang is used for 64-bit builds, while GCC is used for some older 32-bit builds


## Running funfuzz

To run **only the js fuzzers** which compiles shells with random configurations every 8 hours and tests them:

`<python executable> -u funfuzz.loop_bot -b "--random" --target-time 28800 | tee ~/log-loop_botPy.txt`

To test **a patch** (assuming patch is in `~/patch.diff`) against a specific branch (assuming **Mercurial** mozilla-inbound is in `~/trees/mozilla-inbound`), using a debug 64-bit deterministic shell configuration, every 8 hours:

`<python executable> -u funfuzz.loop_bot -b "--enable-debug --enable-more-deterministic -R ~/trees/mozilla-inbound -P ~/patch.diff" --target-time 28800 | tee ~/log-loop_botPy.txt`

In js mode, loop_bot makes use of:

* [compile_shell](js/compile_shell.py)
* [jsfunfuzz](src/funfuzz/js/jsfunfuzz)
* [compare_jit](src/funfuzz/js/compare_jit.py) (if testing deterministic builds)
* randorderfuzz (included in funfuzz, if tests are present in the mozilla repository)
* funbind (Linux-only, included in funfuzz, if [binaryen](https://github.com/WebAssembly/binaryen/releases) can be downloaded)
* [autobisectjs](src/funfuzz/autobisectjs/README.md) (if the mozilla repository is present).

The parameters in `-b` get passed into [compile_shell](js/compile_shell.py) and [autobisectjs](src/funfuzz/autobisectjs/README.md).

You will also need to need a `~/.fuzzmanagerconf` file, similar to:

```
[Main]
serverhost = <your hostname>
serverport = <your port>
serverproto = https
serverauthtoken = <if any>
sigdir = /Users/<your username>/sigcache/
tool = jsfunfuzz
```

Replace anything between `<` and `>` with your desired parameters.

## FAQ:

**Q: What platforms does funfuzz run on?**

**A:** compile_shell has been tested on:

* Windows 10 with [MozillaBuild 3.2](https://wiki.mozilla.org/MozillaBuild)
* macOS 10.13 and 10.14
* Ubuntu 18.04 LTS (only LTS versions supported going forward)

Fedora Linux and openSUSE Leap (42.3 and later) have not been tested extensively and there may be a few bugs along the way.

The following operating systems are less common and while they may still work, be prepared to **expect issues** along the way:

* Windows 7, 8 / Windows 8.1
* Windows Server 2012 R2
* Ubuntu Linux 16.04 LTS (install Python 3.6 via a PPA)
* Ubuntu Linux 15.10 and prior

Support for the following operating systems **have been removed**:

* Windows Vista, Windows XP and earlier
* Mac OS X 10.12 and earlier
* Ubuntu Linux 13.10 and earlier
* Ubuntu (and variants) on [ARM ODROID boards](http://www.hardkernel.com/main/main.php)

**Q: What version of Python does funfuzz require?**

**A:** Python 3.6+


CHANGELOG.md
## 0.6.0 (2019-02-19)

0.6.x supports Python 3.6+ only

Features:

* funbind: Experimental integration with binaryen landed (Linux-only) (#219)
* funbind: binaryen version bumped to 68
* jsfunfuzz: Initial support for mark bit and gray root functions, `newGlobal({newCompartment: true})`, `Object.values`, `Object.[get|set]PrototypeOf` and `enableShapeConsistencyChecks()`
* compare_jit: Ignore `Object.getOwnPropertyNames`, `dumpScopeChain`, `addMarkObservers`, `clearMarkObservers` and `getMarks`
* randorderfuzz: support streams tests
* Use exponential backoff for wasm file execution
* Add a lock using fasteners to prevent `wasm-opt` from tripping over itself
* Use exponential backoff for FuzzManager submission (#145)
* ARM64 code improvements (both simulator and native)
* ICU support bumped up to version 63
* Max gczeal value bumped up to 25
* Bump minimum macOS to be 10.13.6
* Shell builds are compiled with `--disable-cranelift` only if on [m-c rev 6fcf54117a3b](https://hg.mozilla.org/mozilla-central/rev/6fcf54117a3b) or later, till current m-c tip
* Make workaround for compiling further back, on Linux systems with sed >= 4.3 and add tests
* Use GCC for 32-bit builds when bisecting back prior to [m-c rev e1cac03485d9](https://hg.mozilla.org/mozilla-central/rev/e1cac03485d9)
* Support `--more-compartments` in most places
* `--enable-streams` has been deprecated in favour of `--no-streams`
* (all code relating to Python 2.7 support have been removed)

Bugfixes:

* funbind: Disable on ARM64 Linux due to binaryen GH issue 1615
* Fixed TypeError thrown when `file_contains_str` is run after move to Python 3.6+ (#220)
* Do not specify function names in `__init__.py` since we are now on Python 3.6+ to fix RuntimeWarning (#208)
* Remove weights in build_options for slow devices since we do not deal with those anymore
* Remove `--ion-loop-unrolling=on/off` as per [bug 1520998](https://bugzilla.mozilla.org/show_bug.cgi?id=1520998)
* Remove flags related to `--no-wasm` from compare_jit testing, replacing with new ones, e.g. `--wasm-compiler=[none|baseline|ion|baseline+ion]`
* Removed some subprocess calls in favour of the more pythonic way, e.g. for gzipping code
* Remove "-backup" file logic as its support was flaky
* Tweak packaging mode to use find_packages from setuptools in setup.py
* Windows Asan binary support fixes
* Library version bumps
* Various other bugfixes

Testing-related:

* code coverage tests added (#202)
* funfuzz now uses `pytest` throughout, old `unittest`-related code has been removed
* Switch to using `--stream` when running hg clone, on Travis
* `shellcheck` and `bashate` now run on Travis for bash scripts
* Fast tests run on Travis for macOS and Windows
* AppVeyor integration has been removed from funfuzz
* Support pylint 2.x on Python 3 (#218)
* Add cleanup script to wipe `*.pyc`, `*.pyo` files and `__pycache__` dirs, run flake8, fast pytests and pylint
* Various other Travis/testing bugfixes

## 0.5.0 (2018-11-07)

0.5.x is the final version series with stable dual Python 2/3 support, and the branch will then be put on maintenance mode. Going forward, funfuzz will be on Python 3.6+

Features:

* funfuzz: Numerous Python 3 compatibility fixes
* funfuzz: SpiderMonkey code coverage support added
* funfuzz: The `crashesat` interestingness test has been refactored to use argparse, logging and pathlib (#199)
* funfuzz: `flake8-commas`, `flake8-quotes` extensions to `flake8` linting were added
* funfuzz: There is now a `get_hg_repo.sh` script in the util directory to clone `mozilla-central` or `mozilla-beta` using aria2 instead
* funfuzz: Remove `shellify` (#184)
* funfuzz/compare_jit: `--no-streams` and `--enable-wasm-gc` are now tested
* funfuzz/compare_jit: Removed `--ion-shared-stubs=[on|off]`, `--non-writable-jitcode`, `--ion-aa=flow-sensitive` and `--ion-aa=flow-insensitive` since they are no longer part of SpiderMonkey
* jsfunfuzz: `objectEmulatingUndefined` became `createIsHTMLDDA`, see [bug 1410194](https://bugzilla.mozilla.org/show_bug.cgi?id=1410194)
* jsfunfuzz: Generates decreasing for-loops
* jsfunfuzz: Tests the `keepFailing:true` option for oomTest
* jsfunfuzz: Support `recomputeWrappers`
* jsfunfuzz: `evaluate` accepts `saveIncrementalBytecode` as a parameter, see [bug 1427860](https://bugzilla.mozilla.org/show_bug.cgi?id=1427860)
* jsfunfuzz: `newGlobal` accepts `sameCompartmentAs` as a parameter, see [bug 1487238](https://bugzilla.mozilla.org/show_bug.cgi?id=1487238)
* jsfunfuzz: `newGlobal` accepts `invisibleToDebugger` as a parameter
* jsfunfuzz: Object.prototype no longer have the `__count__` and `__parent__` properties
* jsfunfuzz: Stop generating generator expressions

Bugfixes:

* compare_jit: `--no-native-regexp` and `--no-wasm` were removed from basic_flag_sets
* compare_jit: Calling `ShellResult` in `js_interesting` would fail due to the absence of `options.jsengine`
* funfuzz: Fix #9 - compileShell fails on Fedora due to autoconf 2.13 binary name discrepancy (#189)
* funfuzz: Fix #33 - Dump the error to `.busted` log files when configuration fails, but append the info to them if they already exist
* funfuzz: Off-by-one error in path concatenation in `jsFilesIn` function fixed
* funfuzz: Some Clang/ASan build support fixes
* funfuzz: Rename `sps` to `os_ops` in loop.py (#205)
* Various other bugfixes

## 0.4.2 (2018-04-20)

Bugfixes:
*  Fix #185 - `Commandline argument -t "js" in bot.py is not recognized. But it is mentioned in the readme.md.`
*  Disable tests involving compile_shell on Python 2.7 mode unless Python 3.5+ is installed due to a mozilla-central requirement
  * This temporarily lowers code coverage numbers reported to Codecov via Travis, until Python 3.5+ support is finished
* Integrated the isort Python checker into flake8 linting process
* More automation / documentation / linting fixes

Notes:
* Windows is still on the 0.1.x legacy branch, until Python 3.5+ support is completed

## 0.4.1 (2018-04-19)

Bugfixes:

js/shellFlags.py
#!/usr/bin/env python
# coding=utf-8
# pylint: disable=import-error,invalid-name,missing-docstring
# pylint: disable=too-many-branches,too-many-statements,wrong-import-position
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at https://mozilla.org/MPL/2.0/.

from __future__ import absolute_import, print_function

import multiprocessing
import os
import random
import sys

import inspectShell

path0 = os.path.dirname(os.path.abspath(__file__))
path1 = os.path.abspath(os.path.join(path0, os.pardir, 'util'))
sys.path.append(path1)
import subprocesses as sps


def memoize(f, cache=None):
    """Function decorator that caches function results."""
    cache = cache or {}

    # From http://code.activestate.com/recipes/325205-cache-decorator-in-python-24/#c9
    def g(*args, **kwargs):
        key = (f, tuple(args), frozenset(kwargs.items()))
        if key not in cache:
            cache[key] = f(*args, **kwargs)
        return cache[key]
    return g


@memoize
def shellSupportsFlag(shellPath, flag):
    return inspectShell.shellSupports(shellPath, [flag, '-e', '42'])


def chance(p):
    return random.random() < p


def randomFlagSet(shellPath):
    """Return a random list of CLI flags appropriate for the given shell.

    Only works for spidermonkey js shell. Does not work for xpcshell.
    """
    args = []

    ion = shellSupportsFlag(shellPath, "--ion") and chance(.8)

    if shellSupportsFlag(shellPath, '--fuzzing-safe'):
        args.append("--fuzzing-safe")  # --fuzzing-safe landed in bug 885361

    # Landed in m-c changeset c0c1d923c292, see bug 1255008
    if shellSupportsFlag(shellPath, '--ion-aa=flow-sensitive'):
        if chance(.4):
            args.append('--ion-aa=flow-sensitive')
        elif shellSupportsFlag(shellPath, '--ion-aa=flow-insensitive') and chance(.4):
            args.append('--ion-aa=flow-insensitive')

    # See bug 932517, which had landed to fix this issue. Keeping this around for archives:
    #   Original breakage in m-c rev 269359 : https://hg.mozilla.org/mozilla-central/rev/a0ccab2a6e28
    #   Fix in m-c rev 269896: https://hg.mozilla.org/mozilla-central/rev/3bb8446a6d8d
    # Anything in-between involving let probably needs "-e 'version(185);'" to see if we can bypass breakage
    # if shellSupportsFlag(shellPath, "--execute='version(185);'"):
    #     args.append("--execute='version(185);'")
    # Note that the version function was removed in m-c rev 392455 (Fx59) - 589914e65db7

    # Note for future: --wasm-check-bce is only useful for x86 and ARM32

    if shellSupportsFlag(shellPath, '--wasm-always-baseline') and chance(.5):
        args.append("--wasm-always-baseline")  # --wasm-always-baseline landed in bug 1232205

    if shellSupportsFlag(shellPath, '--ion-pgo=on') and chance(.2):
        args.append("--ion-pgo=on")  # --ion-pgo=on landed in bug 1209515

    if shellSupportsFlag(shellPath, '--ion-sincos=on') and chance(.5):
        sincosValue = "on" if chance(0.5) else "off"
        args.append("--ion-sincos=" + sincosValue)  # --ion-sincos=[on|off] landed in bug 984018

    if shellSupportsFlag(shellPath, '--ion-instruction-reordering=on') and chance(.2):
        args.append("--ion-instruction-reordering=on")  # --ion-instruction-reordering=on landed in bug 1195545

    if shellSupportsFlag(shellPath, '--ion-shared-stubs=on') and chance(.2):
        args.append("--ion-shared-stubs=on")  # --ion-shared-stubs=on landed in bug 1168756

    if shellSupportsFlag(shellPath, '--non-writable-jitcode') and chance(.3):
        args.append("--non-writable-jitcode")  # --non-writable-jitcode landed in bug 977805

    if shellSupportsFlag(shellPath, "--execute=setJitCompilerOption('ion.forceinlineCaches',1)") and chance(.1):
        args.append("--execute=setJitCompilerOption('ion.forceinlineCaches',1)")

    if shellSupportsFlag(shellPath, '--no-cgc') and chance(.1):
        args.append("--no-cgc")  # --no-cgc landed in bug 1126769

    if shellSupportsFlag(shellPath, '--no-ggc') and chance(.1):
        args.append("--no-ggc")  # --no-ggc landed in bug 706885

    if shellSupportsFlag(shellPath, '--no-incremental-gc') and chance(.1):
        args.append("--no-incremental-gc")  # --no-incremental-gc landed in bug 958492

    if shellSupportsFlag(shellPath, '--no-unboxed-objects') and chance(.2):
        args.append("--no-unboxed-objects")  # --no-unboxed-objects landed in bug 1162199

    # if shellSupportsFlag(shellPath, '--ion-sink=on') and chance(.2):
    #    args.append("--ion-sink=on")  # --ion-sink=on landed in bug 1093674

    if shellSupportsFlag(shellPath, '--gc-zeal=0') and chance(.9):
        # Focus testing on CheckNursery (16), see:
        #     https://hg.mozilla.org/mozilla-central/rev/bdbb5822afe1
        gczealValue = 16 if chance(0.5) else random.randint(0, 16)
        args.append("--gc-zeal=" + str(gczealValue))  # --gc-zeal= landed in bug 1101602

    if shellSupportsFlag(shellPath, '--enable-small-chunk-size') and chance(.1):
        args.append("--enable-small-chunk-size")  # --enable-small-chunk-size landed in bug 941804

    if shellSupportsFlag(shellPath, '--ion-loop-unrolling=on') and chance(.2):
        args.append("--ion-loop-unrolling=on")  # --ion-loop-unrolling=on landed in bug 1039458

    if shellSupportsFlag(shellPath, '--no-threads') and chance(.5):
        args.append("--no-threads")  # --no-threads landed in bug 1031529

    if shellSupportsFlag(shellPath, '--disable-ion') and chance(.05):
        args.append("--disable-ion")  # --disable-ion landed in bug 789319

    # See bug 1026919 comment 60:
    if sps.isARMv7l and \
            shellSupportsFlag(shellPath, '--arm-asm-nop-fill=0') and chance(0.3):
        # It was suggested to focus more on the range between 0 and 1.
        # Reduced the upper limit to 8, see bug 1053996 comment 8.
        asmNopFill = random.randint(1, 8) if chance(0.3) else random.randint(0, 1)
        args.append("--arm-asm-nop-fill=" + str(asmNopFill))  # Landed in bug 1020834

    # See bug 1026919 comment 60:
    if sps.isARMv7l and \
            shellSupportsFlag(shellPath, '--asm-pool-max-offset=1024') and chance(0.3):
        asmPoolMaxOffset = random.randint(5, 1024)
        args.append("--asm-pool-max-offset=" + str(asmPoolMaxOffset))  # Landed in bug 1026919

    if shellSupportsFlag(shellPath, '--no-native-regexp') and chance(.1):
        args.append("--no-native-regexp")  # See bug 976446

    if inspectShell.queryBuildConfiguration(shellPath, 'arm-simulator') and chance(.4):
        args.append('--arm-sim-icache-checks')

    if (shellSupportsFlag(shellPath, '--no-sse3') and shellSupportsFlag(shellPath, '--no-sse4')) and chance(.2):
        # --no-sse3 and --no-sse4 landed in m-c rev 526ba3ace37a.
        if chance(.5):
            args.append("--no-sse3")
        else:
            args.append("--no-sse4")

    # We should stop fuzzing --no-fpu according to the js devs...
    # if shellSupportsFlag(shellPath, '--no-fpu') and chance(.2):
    #     args.append("--no-fpu")  # --no-fpu landed in bug 858022

    if shellSupportsFlag(shellPath, '--no-asmjs') and chance(.5):
        args.append("--no-asmjs")

    # --baseline-eager landed after --no-baseline on the IonMonkey branch prior to landing on m-c.
    if shellSupportsFlag(shellPath, '--baseline-eager'):
        if chance(.3):
            args.append('--no-baseline')
        # elif is important, as we want to call --baseline-eager only if --no-baseline is not set.
        elif chance(.6):
            args.append("--baseline-eager")

    if shellSupportsFlag(shellPath, '--ion-offthread-compile=off'):
        if chance(.7):
            # Focus on the reproducible cases
            args.append("--ion-offthread-compile=off")
        elif chance(.5) and multiprocessing.cpu_count() > 1 and \
                shellSupportsFlag(shellPath, '--thread-count=1'):
            # Adjusts default number of threads for parallel compilation (turned on by default)
            totalThreads = random.randint(2, (multiprocessing.cpu_count() * 2))
            args.append('--thread-count=' + str(totalThreads))
        # else:
        #   Default is to have --ion-offthread-compile=on and --thread-count=<some default value>
    elif shellSupportsFlag(shellPath, '--ion-parallel-compile=off'):
        # --ion-parallel-compile=off has gone away as of m-c rev 9ab3b097f304 and f0d67b1ccff9.
        if chance(.7):
            # Focus on the reproducible cases
            args.append("--ion-parallel-compile=off")
        elif chance(.5) and multiprocessing.cpu_count() > 1 and \
                shellSupportsFlag(shellPath, '--thread-count=1'):
            # Adjusts default number of threads for parallel compilation (turned on by default)
            totalThreads = random.randint(2, (multiprocessing.cpu_count() * 2))
            args.append('--thread-count=' + str(totalThreads))
        # else:
        #   The default is to have --ion-parallel-compile=on and --thread-count=<some default value>

    if ion:
        if chance(.6):
            args.append("--ion-eager")
        if chance(.2):
            args.append("--ion-gvn=off")
        if chance(.2):
            args.append("--ion-licm=off")
        if shellSupportsFlag(shellPath, '--ion-edgecase-analysis=off') and chance(.2):
            args.append("--ion-edgecase-analysis=off")
        if chance(.2):
            args.append("--ion-range-analysis=off")
        if chance(.2):
            args.append("--ion-inlining=off")
        if chance(.2):
            args.append("--ion-osr=off")
        if chance(.2):
            args.append("--ion-limit-script-size=off")
        # Backtracking (on by default as of 2015-04-15) and stupid landed in m-c changeset dc4887f61d2e
        # The stupid allocator isn't used by default and devs prefer not to have to fix fuzzbugs
        # if shellSupportsFlag(shellPath, '--ion-regalloc=stupid') and chance(.2):
            # args.append('--ion-regalloc=stupid')
        if shellSupportsFlag(shellPath, '--ion-regalloc=testbed') and chance(.2):
            args.append('--ion-regalloc=testbed')
        if shellSupportsFlag(shellPath, '--ion-check-range-analysis') and chance(.3):
            args.append('--ion-check-range-analysis')
        if shellSupportsFlag(shellPath, '--ion-extra-checks') and chance(.3):
            args.append('--ion-extra-checks')
    else:
        args.append("--no-ion")

    # if chance(.05):
    #    args.append("--execute=verifyprebarriers()")

    if chance(.05):
        args.append("-D")  # aka --dump-bytecode

    return args


def basicFlagSets(shellPath):
    """These flag combos are used w/the original flag sets when run through Lithium & autoBisect."""
    if shellSupportsFlag(shellPath, "--no-threads"):
        basicFlagList = [
            # Parts of this flag permutation come from:
            # https://hg.mozilla.org/mozilla-central/file/c91249f41e37/js/src/tests/lib/tests.py#l13
            # compareJIT uses the following first flag set as the sole baseline when fuzzing
            ['--fuzzing-safe', '--no-threads', '--ion-eager'],
            ['--fuzzing-safe', '--ion-offthread-compile=off', '--ion-eager'],
            ['--fuzzing-safe', '--ion-offthread-compile=off'],
            ['--fuzzing-safe', '--baseline-eager'],
            # ['--fuzzing-safe', '--baseline-eager', '--no-fpu'],  # --no-fpu seems to be deprecated now
            ['--fuzzing-safe', '--no-baseline', '--no-ion'],
        ]
        if shellSupportsFlag(shellPath, "--non-writable-jitcode"):
            basicFlagList.append(['--fuzzing-safe', '--no-threads', '--ion-eager',
                                  '--non-writable-jitcode', '--ion-check-range-analysis',
                                  '--ion-extra-checks', '--no-sse3'])
        if shellSupportsFlag(shellPath, "--no-wasm"):
            basicFlagList.append(['--fuzzing-safe', '--no-baseline', '--no-asmjs',
                                  '--no-wasm', '--no-native-regexp'])
        if shellSupportsFlag(shellPath, "--wasm-always-baseline"):
            basicFlagList.append(['--fuzzing-safe', '--no-threads', '--ion-eager',
                                  '--wasm-always-baseline'])
        return basicFlagList
    elif shellSupportsFlag(shellPath, "--ion-offthread-compile=off"):
        basicFlagList = [
            # Parts of this flag permutation come from:
            # https://hg.mozilla.org/mozilla-central/file/84bd8d9f4256/js/src/tests/lib/tests.py#l12
            # as well as other interesting flag combinations that have found / may find new bugs.
            # compareJIT uses the following first flag set as the sole baseline when fuzzing
            ['--fuzzing-safe', '--ion-offthread-compile=off'],
            ['--fuzzing-safe', '--ion-offthread-compile=off', '--no-baseline'],  # Not in jit_test.py though...
            ['--fuzzing-safe', '--ion-offthread-compile=off', '--no-baseline', '--no-ion'],
            ['--fuzzing-safe', '--ion-offthread-compile=off', '--no-baseline', '--ion-eager'],  # Not in jit_test.py...
            ['--fuzzing-safe', '--ion-offthread-compile=off', '--ion-eager'],  # Not in jit_test.py though...
            # This is not in jit_test.py, see bug 848906 comment 1
            ['--fuzzing-safe', '--ion-offthread-compile=off', '--no-ion'],
            # ['--fuzzing-safe', '--ion-offthread-compile=off', '--no-fpu'],  # --no-fpu seems to be deprecated now
        ]
        if shellSupportsFlag(shellPath, "--thread-count=1"):
            basicFlagList.append(['--fuzzing-safe', '--ion-offthread-compile=off', '--ion-eager'])
            # Range analysis had only started to stabilize around the time when --no-sse3 landed.
            if shellSupportsFlag(shellPath, '--no-sse3'):
                basicFlagList.append(['--fuzzing-safe', '--ion-offthread-compile=off',
                                      '--ion-eager', '--ion-check-range-analysis', '--no-sse3'])
        return basicFlagList
    else:
        basicFlagList = [
            # Parts of this flag permutation come from:
            # https://hg.mozilla.org/mozilla-central/file/10932f3a0ba0/js/src/tests/lib/tests.py#l12
            # as well as other interesting flag combinations that have found / may find new bugs.
            # compareJIT uses the following first flag set as the sole baseline when fuzzing
            ['--fuzzing-safe', '--ion-parallel-compile=off'],
            ['--fuzzing-safe', '--ion-parallel-compile=off', '--no-baseline'],  # Not in jit_test.py though...
            ['--fuzzing-safe', '--ion-parallel-compile=off', '--no-baseline', '--no-ion'],
            ['--fuzzing-safe', '--ion-parallel-compile=off', '--no-baseline', '--ion-eager'],  # Not in jit_test.py...
            ['--fuzzing-safe', '--ion-parallel-compile=off', '--ion-eager'],  # Not in jit_test.py though...
            ['--fuzzing-safe', '--ion-parallel-compile=off', '--baseline-eager'],
            # See bug 848906 comment 1
            ['--fuzzing-safe', '--ion-parallel-compile=off', '--baseline-eager', '--no-ion'],
            # --no-fpu seems to be deprecated now
            # ['--fuzzing-safe', '--ion-parallel-compile=off', '--baseline-eager', '--no-fpu'],
        ]
        if shellSupportsFlag(shellPath, "--thread-count=1"):
            basicFlagList.append(['--fuzzing-safe', '--ion-eager', '--ion-parallel-compile=off'])
            # Range analysis had only started to stabilize around the time when --no-sse3 landed.
            if shellSupportsFlag(shellPath, '--no-sse3'):
                basicFlagList.append(['--fuzzing-safe', '--ion-parallel-compile=off',
                                      '--ion-eager', '--ion-check-range-analysis', '--no-sse3'])
        return basicFlagList


# Consider adding a function (for compareJIT reduction) that takes a flag set
# and returns all its (meaningful) subsets.


def testRandomFlags():
    for _ in range(100):
        print(" ".join(randomFlagSet(sys.argv[1])))


if __name__ == "__main__":
    testRandomFlags()
